File: src/main/java/com/refinedmods/refinedstorage/network/FilterUpdateMessage.java
package com.refinedmods.refinedstorage.network;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.container.FilterContainerMenu;
import com.refinedmods.refinedstorage.item.FilterItem;
import com.refinedmods.refinedstorage.util.PacketBufferUtils;

import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import net.neoforged.neoforge.network.handling.PlayPayloadContext;

public class FilterUpdateMessage implements CustomPacketPayload {
    public static final ResourceLocation ID = new ResourceLocation(RS.ID, "filter_update");

    private final int compare;
    private final int mode;
    private final boolean modFilter;
    private final String name;
    private final int type;

    public FilterUpdateMessage(int compare, int mode, boolean modFilter, String name, int type) {
        this.compare = compare;
        this.mode = mode;
        this.modFilter = modFilter;
        this.name = name;
        this.type = type;
    }

    public static FilterUpdateMessage decode(FriendlyByteBuf buf) {
        return new FilterUpdateMessage(
            buf.readInt(),
            buf.readInt(),
            buf.readBoolean(),
            PacketBufferUtils.readString(buf),
            buf.readInt()
        );
    }

    public static void handle(FilterUpdateMessage message, PlayPayloadContext ctx) {
        ctx.player().ifPresent(player -> ctx.workHandler().submitAsync(() -> {
            FilterItem.setCompare(((FilterContainerMenu) player.containerMenu).getFilterItem(), message.compare);
            FilterItem.setMode(((FilterContainerMenu) player.containerMenu).getFilterItem(), message.mode);
            FilterItem.setModFilter(((FilterContainerMenu) player.containerMenu).getFilterItem(), message.modFilter);
            FilterItem.setName(((FilterContainerMenu) player.containerMenu).getFilterItem(), message.name);
            FilterItem.setType(((FilterContainerMenu) player.containerMenu).getFilterItem(), message.type);
        }));
    }

    @Override
    public void write(FriendlyByteBuf buf) {
        buf.writeInt(compare);
        buf.writeInt(mode);
        buf.writeBoolean(modFilter);
        buf.writeUtf(name);
        buf.writeInt(type);
    }

    @Override
    public ResourceLocation id() {
        return ID;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/network/FilterUpdateMessage.kt
package com.refinedmods.refinedstorage.network

import com.refinedmods.refinedstorage.container.FilterContainer
import com.refinedmods.refinedstorage.item.FilterItem.Companion.setCompare
import com.refinedmods.refinedstorage.item.FilterItem.Companion.setModFilter
import com.refinedmods.refinedstorage.item.FilterItem.Companion.setMode
import com.refinedmods.refinedstorage.item.FilterItem.Companion.setName
import com.refinedmods.refinedstorage.item.FilterItem.Companion.setType
import com.refinedmods.refinedstorage.util.PacketByteBufUtils
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.network.PacketByteBuf
import net.minecraftforge.fml.network.NetworkEvent
import java.util.function.Supplier

class FilterUpdateMessage(private val compare: Int, private val mode: Int, private val modFilter: Boolean, private val name: String, private val type: Int) {
    companion object {
        fun decode(buf: PacketByteBuf): FilterUpdateMessage {
            return FilterUpdateMessage(
                    buf.readInt(),
                    buf.readInt(),
                    buf.readBoolean(),
                    PacketByteBufUtils.readString(buf),
                    buf.readInt()
            )
        }

        fun encode(message: FilterUpdateMessage, buf: PacketByteBuf) {
            buf.writeInt(message.compare)
            buf.writeInt(message.mode)
            buf.writeBoolean(message.modFilter)
            buf.writeString(message.name)
            buf.writeInt(message.type)
        }

        fun handle(message: FilterUpdateMessage, ctx: Supplier<NetworkEvent.Context>) {
            val player: PlayerEntity = ctx.get().getSender()
            if (player != null && player.openContainer is FilterContainer) {
                ctx.get().enqueueWork({
                    setCompare((player.openContainer as FilterContainer).stack, message.compare)
                    setMode((player.openContainer as FilterContainer).stack, message.mode)
                    setModFilter((player.openContainer as FilterContainer).stack, message.modFilter)
                    setName((player.openContainer as FilterContainer).stack, message.name)
                    setType((player.openContainer as FilterContainer).stack, message.type)
                })
            }
            ctx.get().setPacketHandled(true)
        }
    }
}